dynamodb/batch/batch.go

```diff
@@ -0,0 +1,150 @@
+	requestLimit = 25
+	// After the sleep reaches this value of seconds (for retrying unprocessed items), we error with a time-out
+	maxTimeout int64 = 10
+)
+
+// Batch takes on the job of tracking BatchWrite state
+type Batch struct {
+	table        string
+	bwInput      dynamodb.BatchWriteItemInput
+	requests     []*dynamodb.WriteRequest
+	requestLimit int   // number of requests per batch
+	maxTimeout   int64 // maximum time to sleep between retries
+	dynamoapi    dynamodbiface.DynamoDBAPI
+	sleepSeconds int64 // Backoff
+}
+
+// Iface presents a test interface for Batch
+type Iface interface {
+	SetTableName(tableName string)
+	SetRequestLimit(newLimit int) error
+	SetMaxTimeout(newTimeout int64) error
+	Add(requestType RequestType, item map[string]*dynamodb.AttributeValue) ([]*dynamodb.ConsumedCapacity, error)
+	Send() ([]*dynamodb.ConsumedCapacity, error)
+}
+
+// NewBatch returns a new object with the given table name, request limit, and dynamo API
+func NewBatch(ddb dynamodbiface.DynamoDBAPI) *Batch {
+	return &Batch{
+		dynamoapi:    ddb,
+		sleepSeconds: 1,
+		requestLimit: requestLimit,
+		maxTimeout:   maxTimeout,
+		bwInput:      basicBatchInput(),
+	}
+}
+
+func basicBatchInput() dynamodb.BatchWriteItemInput {
+	return dynamodb.BatchWriteItemInput{
+		RequestItems:           map[string][]*dynamodb.WriteRequest{},
+		ReturnConsumedCapacity: aws.String("TOTAL"),
+	}
+}
+
+// SetTableName sets the table name
+func (b *Batch) SetTableName(tableName string) {
+	b.table = tableName
+	return
+}
+
+// SetRequestLimit sets the number of items to pile up before sending
+func (b *Batch) SetRequestLimit(newLimit int) error {
+	if newLimit > requestLimit {
+		return fmt.Errorf("requestLimit must be %d or less", requestLimit)
+	}
+	if newLimit < 1 {
+		return fmt.Errorf("requestLimit must be %d or less", requestLimit)
+	}
+	b.requestLimit = newLimit
+	return nil
+}
+
+// SetMaxTimeout sets the maximum sleep allowed for retries
+func (b *Batch) SetMaxTimeout(newTimeout int64) error {
+	if newTimeout > maxTimeout {
+		return fmt.Errorf("too much timeout, should be less than %d", maxTimeout)
+	}
+	if newTimeout < 1 {
+		return fmt.Errorf("not enough time out, should be at least 1")
+	}
+	b.maxTimeout = int64(newTimeout)
+	return nil
+}
+
+// Add an item to the batch, and send the batch if it's full
+func (b *Batch) Add(requestType RequestType, item map[string]*dynamodb.AttributeValue) ([]*dynamodb.ConsumedCapacity, error) {
+	if b.table == "" {
+		return nil, fmt.Errorf("table name required, call SetTableName")
+	}
+	request := &dynamodb.WriteRequest{}
+	switch requestType {
+	case PutRequest:
+		request.PutRequest = &dynamodb.PutRequest{Item: item}
+	case DeleteRequest:
+		request.DeleteRequest = &dynamodb.DeleteRequest{Key: item}
+	}
+	b.requests = append(b.requests, request)
+	if len(b.requests) >= b.requestLimit {
+		return b.Send()
+	}
```

Comment:
This seems counter intuitive. The method is called `Add`, but it sends the request. This behavior is odd. You should always require `Send` be called prior to anything being sent. It is odd behavior to do this automatically but then require an explicit call for the remainder. Of course, when `Send` is called it should send them in batch sizes as expected, including the remainder.

Reply:
My intent here is that the user of this aide sits in a loop, `Add()`ing items as it goes, and after the loop is done, does a `Send()` to "finalize".

Meanwhile, the aide is building up dynamo batches and sending them to AWS when each batch has enough items.

Sort of like how one might print a bunch of lines... which get buffered to the screen and actually displayed whenever. But one might also use an explicit `flush` to display whatever is there.

I can see why that might be confusing. I'll take suggestions on better names.